json: update json doc #11180 (#10389)
pingcap/docs: data-type-json.md

```

For more information, see [JSON Functions](/functions-and-operators/json-functions.md) and [Generated Columns](/generated-columns.md).

## Restrictions

- Currently, TiDB does not support pushing down `JSON` functions to TiFlash.
- TiDB does not support using the range selection syntax in JSON PATH. For example, executing the following SQL statements in TiDB will return errors.

    ```sql
    SELECT j->'$[1 to 2]' FROM t;
    SELECT j->'$[last]' FROM t;
    ```

> - Backup & Restore (BR) versions earlier than v6.3.0 do not support recovering data containing JSON columns. No version of BR supports recovering data containing JSON columns to TiDB clusters earlier than v6.3.0.

Author: ```suggestion
- TiDB Backup & Restore (BR) versions earlier than v6.3.0 do not support recovering data containing JSON columns. No version of BR supports recovering data containing JSON columns to TiDB clusters earlier than v6.3.0.
```

- Do not use any replication tool to replicate data containing non-standard `JSON` data types, such as `DATE`, `DATETIME`, and `TIME`.

## MySQL compatibility

> - When creating JSON columns with data in the `BINARY` type, MySQL mislabels it as the `STRING` type currently.

Reviewer: ```suggestion
- When you create JSON columns with data in the `BINARY` type, MySQL mislabels it as the `STRING` type currently.
```

    1 row in set (0.01 sec)
    ```

    For more information, see issue [#37443](https://github.com/pingcap/tidb/issues/37443).

- When converting the data type from `ENUM` or `SET` to `JSON`, TiDB checks the correctness of data format. For example, executing the following SQL statements in TiDB will return an error.

    ```sql
    CREATE TABLE t(e ENUM('a'));
    INSERT INTO t VALUES ('a');
    mysql> SELECT CAST(e AS JSON) FROM t;
    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
    ```

    For more information, see issue [#9999](https://github.com/pingcap/tidb/issues/9999).

- In TiDB, you can use `ORDER BY` to sort JSON arrays or JSON objects.

    In MySQL, if you use `ORDER BY` to sort JSON arrays or JSON objects，MySQL returns a warning and the sorting result does not match the result of the comparison operation:

    ```sql
    CREATE TABLE t(j JSON);
    INSERT INTO t VALUES ('[1,2,3,4]');
    INSERT INTO t VALUES ('[5]');

    mysql> SELECT j FROM t WHERE j < JSON_ARRAY(5);
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    +--------------+
    1 row in set (0.00 sec)

>     -- In TiDB, executing the following SQL statement returns the correct sorting result. In MySQL, executing the following SQL statement returns the "This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. " warning and the sorting result is inconsistent with the comparison result of `<`.

Author: ```suggestion
    -- In TiDB, executing the following SQL statement returns the correct sorting result. In MySQL, executing the following SQL statement returns the "This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'." warning and the sorting result is inconsistent with the comparison result of `<`.
```

    mysql> SELECT j FROM t ORDER BY j;
    +--------------+
    | j            |
    +--------------+
    | [1, 2, 3, 4] |
    | [5]          |
    +--------------+
    2 rows in set (0.00 sec)
    ```

    For more information, see issue [#37506](https://github.com/pingcap/tidb/issues/37506).

- When you insert data to a JSON column, TiDB implicitly converts the value of the data to the `JSON` type.

    ```sql
    CREATE TABLE t(col JSON);

>     -- In TiDB, the following INSERT statement is executed successfully. In MySQL, executing the following INSERT statement returns the `Invalid JSON text` error.

Author: ```suggestion
    -- In TiDB, the following INSERT statement is executed successfully. In MySQL, executing the following INSERT statement returns the "Invalid JSON text" error.
```
